File: code/backend/java/adapter-rest/src/test/java/games/paths/adapters/rest/controller/match/CharacterControllerTest.java
package games.paths.adapters.rest.controller.match;

import games.paths.core.model.match.CharacterInstanceInfo;
import games.paths.core.port.match.CharacterCommandPort;
import games.paths.core.port.match.CharacterCommandPort.CharacterJoinException;
import games.paths.core.port.match.CharacterQueryPort;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.http.MediaType;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.request.MockHttpServletRequestBuilder;
import org.springframework.test.web.servlet.setup.MockMvcBuilders;

import java.util.List;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.*;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;

class CharacterControllerTest {

    private MockMvc mockMvc;
    private CharacterCommandPort commandPort;
    private CharacterQueryPort queryPort;

    @BeforeEach
    void setUp() {
        commandPort = mock(CharacterCommandPort.class);
        queryPort = mock(CharacterQueryPort.class);
        mockMvc = MockMvcBuilders.standaloneSetup(new CharacterController(commandPort, queryPort)).build();
    }

    private MockHttpServletRequestBuilder authed(MockHttpServletRequestBuilder b) {
        return b.requestAttr("userUuid", "user-uuid");
    }

    private CharacterInstanceInfo info() {
        CharacterInstanceInfo i = new CharacterInstanceInfo();
        i.setUuid("char-uuid");
        i.setMatchUuid("match-uuid");
        i.setUserUuid("user-uuid");
        i.setDexterity(19);
        i.setLife(137);
        i.setTraitUuids(List.of("trait-1"));
        return i;
    }

    // ─── join ────────────────────────────────────────────────────────────────

    @Test
    void join_unauthenticated_401() throws Exception {
        mockMvc.perform(post("/api/matches/m1/join")
                        .contentType(MediaType.APPLICATION_JSON).content("{}"))
                .andExpect(status().isUnauthorized());
    }

    @Test
    void join_success_201() throws Exception {
        when(commandPort.join(any())).thenReturn(info());
        mockMvc.perform(authed(post("/api/matches/m1/join"))
                        .contentType(MediaType.APPLICATION_JSON)
                        .content("{\"characterTemplateUuid\":\"t\",\"classUuid\":\"c\",\"traitUuids\":[\"x\"]}"))
                .andExpect(status().isCreated())
                .andExpect(jsonPath("$.uuid").value("char-uuid"))
                .andExpect(jsonPath("$.life").value(137));
    }

    @Test
    void join_emptyBody_usesStoredLoadout_201() throws Exception {
        when(commandPort.join(any())).thenReturn(info());
        mockMvc.perform(authed(post("/api/matches/m1/join")))
                .andExpect(status().isCreated());
    }

    @Test
    void join_matchNotFound_404() throws Exception {
        when(commandPort.join(any())).thenThrow(
                new CharacterJoinException(CharacterJoinException.Code.MATCH_NOT_FOUND, "x"));
        mockMvc.perform(authed(post("/api/matches/m1/join"))
                        .contentType(MediaType.APPLICATION_JSON).content("{}"))
                .andExpect(status().isNotFound());
    }

    @Test
    void join_templateNotFound_404() throws Exception {
        when(commandPort.join(any())).thenThrow(
                new CharacterJoinException(CharacterJoinException.Code.TEMPLATE_NOT_FOUND, "x"));
        mockMvc.perform(authed(post("/api/matches/m1/join"))
                        .contentType(MediaType.APPLICATION_JSON).content("{}"))
                .andExpect(status().isNotFound());
    }

    @Test
    void join_classNotFound_404() throws Exception {
        when(commandPort.join(any())).thenThrow(
                new CharacterJoinException(CharacterJoinException.Code.CLASS_NOT_FOUND, "x"));
        mockMvc.perform(authed(post("/api/matches/m1/join"))
                        .contentType(MediaType.APPLICATION_JSON).content("{}"))
                .andExpect(status().isNotFound());
    }

    @Test
    void join_userNotFound_404() throws Exception {
        when(commandPort.join(any())).thenThrow(
                new CharacterJoinException(CharacterJoinException.Code.USER_NOT_FOUND, "x"));
        mockMvc.perform(authed(post("/api/matches/m1/join"))
                        .contentType(MediaType.APPLICATION_JSON).content("{}"))
                .andExpect(status().isNotFound());
    }

    @Test
    void join_userBanned_403() throws Exception {
        when(commandPort.join(any())).thenThrow(
                new CharacterJoinException(CharacterJoinException.Code.USER_BANNED, "x"));
        mockMvc.perform(authed(post("/api/matches/m1/join"))
                        .contentType(MediaType.APPLICATION_JSON).content("{}"))
                .andExpect(status().isForbidden());
    }

    @Test
    void join_alreadyJoined_409() throws Exception {
        when(commandPort.join(any())).thenThrow(
                new CharacterJoinException(CharacterJoinException.Code.ALREADY_JOINED, "x"));
        mockMvc.perform(authed(post("/api/matches/m1/join"))
                        .contentType(MediaType.APPLICATION_JSON).content("{}"))
                .andExpect(status().isConflict());
    }

    @Test
    void join_classNotCompatible_409() throws Exception {
        when(commandPort.join(any())).thenThrow(
                new CharacterJoinException(CharacterJoinException.Code.CLASS_NOT_COMPATIBLE, "x"));
        mockMvc.perform(authed(post("/api/matches/m1/join"))
                        .contentType(MediaType.APPLICATION_JSON).content("{}"))
                .andExpect(status().isConflict());
    }

    @Test
    void join_matchNotJoinable_409() throws Exception {
        when(commandPort.join(any())).thenThrow(
                new CharacterJoinException(CharacterJoinException.Code.MATCH_NOT_JOINABLE, "x"));
        mockMvc.perform(authed(post("/api/matches/m1/join"))
                        .contentType(MediaType.APPLICATION_JSON).content("{}"))
                .andExpect(status().isConflict());
    }

    @Test
    void join_invalidInput_400() throws Exception {
        when(commandPort.join(any())).thenThrow(
                new CharacterJoinException(CharacterJoinException.Code.INVALID_INPUT, "x"));
        mockMvc.perform(authed(post("/api/matches/m1/join"))
                        .contentType(MediaType.APPLICATION_JSON).content("{}"))
                .andExpect(status().isBadRequest());
    }

    // ─── players ──────────────────────────────────────────────────────────────

    @Test
    void listPlayers_unauthenticated_401() throws Exception {
        mockMvc.perform(get("/api/match/m1/players"))
                .andExpect(status().isUnauthorized());
    }

    @Test
    void listPlayers_ok_200() throws Exception {
        when(queryPort.listPlayers("m1", "user-uuid")).thenReturn(List.of(info()));
        mockMvc.perform(authed(get("/api/match/m1/players")))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$[0].uuid").value("char-uuid"));
    }

    @Test
    void listPlayers_null_404() throws Exception {
        when(queryPort.listPlayers("m1", "user-uuid")).thenReturn(null);
        mockMvc.perform(authed(get("/api/match/m1/players")))
                .andExpect(status().isNotFound());
    }

    // ─── character detail ──────────────────────────────────────────────────────

    @Test
    void getCharacter_unauthenticated_401() throws Exception {
        mockMvc.perform(get("/api/match/m1/characters/c1"))
                .andExpect(status().isUnauthorized());
    }

    @Test
    void getCharacter_ok_200() throws Exception {
        when(queryPort.getCharacter("m1", "c1", "user-uuid")).thenReturn(info());
        mockMvc.perform(authed(get("/api/match/m1/characters/c1")))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.uuid").value("char-uuid"));
    }

    @Test
    void getCharacter_null_404() throws Exception {
        when(queryPort.getCharacter("m1", "c1", "user-uuid")).thenReturn(null);
        mockMvc.perform(authed(get("/api/match/m1/characters/c1")))
                .andExpect(status().isNotFound());
    }
}

File: code/backend/php/tests/Unit/Adapter/Rest/Matches/CharacterControllerTest.php
<?php

declare(strict_types=1);

namespace Games\Paths\Tests\Unit\Adapter\Rest\Matches;

use Games\Paths\Adapter\Rest\Matches\CharacterController;
use Games\Paths\Core\Domain\Matches\CharacterInstanceInfo;
use Games\Paths\Core\Domain\Matches\CharacterJoinException;
use Games\Paths\Core\Port\Matches\CharacterCommandPort;
use Games\Paths\Core\Port\Matches\CharacterQueryPort;
use PHPUnit\Framework\TestCase;
use Slim\Psr7\Factory\ResponseFactory;
use Slim\Psr7\Factory\ServerRequestFactory;

class CharacterControllerTest extends TestCase
{
    private $commandPort;
    private $queryPort;
    private CharacterController $controller;
    private ResponseFactory $responseFactory;
    private ServerRequestFactory $requestFactory;

    protected function setUp(): void
    {
        $this->commandPort = $this->createMock(CharacterCommandPort::class);
        $this->queryPort = $this->createMock(CharacterQueryPort::class);
        $this->controller = new CharacterController($this->commandPort, $this->queryPort);
        $this->responseFactory = new ResponseFactory();
        $this->requestFactory = new ServerRequestFactory();
    }

    private function info(): CharacterInstanceInfo
    {
        return new CharacterInstanceInfo(
            uuid: 'char-uuid', matchUuid: 'match-uuid', userUuid: 'user-uuid',
            characterTemplateUuid: 'tpl', classUuid: 'cls',
            dexterity: 19, intelligence: 18, constitution: 19, energy: 127, life: 137, sad: 0,
            idLocation: 90001, locationUuid: 'loc', locationName: 'location-90001',
            isSleeping: 0, isComa: 0, traitUuids: ['t1'], food: 0, magic: 0, coin: 0
        );
    }

    private function authed(string $method, string $path, array $body = null)
    {
        $req = $this->requestFactory->createServerRequest($method, $path)
            ->withAttribute('userUuid', 'user-uuid');
        if ($body !== null) {
            $req = $req->withParsedBody($body);
        }
        return $req;
    }

    private function resp()
    {
        return $this->responseFactory->createResponse();
    }

    public function testJoinUnauthenticated(): void
    {
        $req = $this->requestFactory->createServerRequest('POST', '/api/matches/m1/join');
        $result = $this->controller->join($req, $this->resp(), ['uuidMatch' => 'm1']);
        $this->assertSame(401, $result->getStatusCode());
    }

    public function testJoinSuccess(): void
    {
        $this->commandPort->method('join')->willReturn($this->info());
        $result = $this->controller->join(
            $this->authed('POST', '/api/matches/m1/join', ['characterTemplateUuid' => 't']),
            $this->resp(), ['uuidMatch' => 'm1']
        );
        $this->assertSame(201, $result->getStatusCode());
        $body = json_decode((string)$result->getBody(), true);
        $this->assertSame('char-uuid', $body['uuid']);
        $this->assertSame(137, $body['life']);
    }

    public function testJoinEmptyBody(): void
    {
        $this->commandPort->method('join')->willReturn($this->info());
        $result = $this->controller->join(
            $this->authed('POST', '/api/matches/m1/join'), $this->resp(), ['uuidMatch' => 'm1']
        );
        $this->assertSame(201, $result->getStatusCode());
    }

    /**
     * @dataProvider errorCodes
     */
    public function testJoinErrorCodes(string $code, int $expected): void
    {
        $this->commandPort->method('join')->willThrowException(new CharacterJoinException($code, 'x'));
        $result = $this->controller->join(
            $this->authed('POST', '/api/matches/m1/join', []), $this->resp(), ['uuidMatch' => 'm1']
        );
        $this->assertSame($expected, $result->getStatusCode());
    }

    public static function errorCodes(): array
    {
        return [
            [CharacterJoinException::MATCH_NOT_FOUND, 404],
            [CharacterJoinException::TEMPLATE_NOT_FOUND, 404],
            [CharacterJoinException::CLASS_NOT_FOUND, 404],
            [CharacterJoinException::USER_NOT_FOUND, 404],
            [CharacterJoinException::USER_BANNED, 403],
            [CharacterJoinException::ALREADY_JOINED, 409],
            [CharacterJoinException::CLASS_NOT_COMPATIBLE, 409],
            [CharacterJoinException::MATCH_NOT_JOINABLE, 409],
            [CharacterJoinException::INVALID_INPUT, 400],
        ];
    }

    public function testPlayersUnauthenticated(): void
    {
        $req = $this->requestFactory->createServerRequest('GET', '/api/match/m1/players');
        $result = $this->controller->listPlayers($req, $this->resp(), ['uuidMatch' => 'm1']);
        $this->assertSame(401, $result->getStatusCode());
    }

    public function testPlayersOk(): void
    {
        $this->queryPort->method('listPlayers')->willReturn([$this->info()]);
        $result = $this->controller->listPlayers(
            $this->authed('GET', '/api/match/m1/players'), $this->resp(), ['uuidMatch' => 'm1']
        );
        $this->assertSame(200, $result->getStatusCode());
        $body = json_decode((string)$result->getBody(), true);
        $this->assertSame('char-uuid', $body[0]['uuid']);
    }

    public function testPlayersNotFound(): void
    {
        $this->queryPort->method('listPlayers')->willReturn(null);
        $result = $this->controller->listPlayers(
            $this->authed('GET', '/api/match/m1/players'), $this->resp(), ['uuidMatch' => 'm1']
        );
        $this->assertSame(404, $result->getStatusCode());
    }

    public function testCharacterUnauthenticated(): void
    {
        $req = $this->requestFactory->createServerRequest('GET', '/api/match/m1/characters/c1');
        $result = $this->controller->getCharacter($req, $this->resp(), ['uuidMatch' => 'm1', 'uuidCharacter' => 'c1']);
        $this->assertSame(401, $result->getStatusCode());
    }

    public function testCharacterOk(): void
    {
        $this->queryPort->method('getCharacter')->willReturn($this->info());
        $result = $this->controller->getCharacter(
            $this->authed('GET', '/api/match/m1/characters/c1'), $this->resp(),
            ['uuidMatch' => 'm1', 'uuidCharacter' => 'c1']
        );
        $this->assertSame(200, $result->getStatusCode());
        $body = json_decode((string)$result->getBody(), true);
        $this->assertSame(['t1'], $body['traitUuids']);
    }

    public function testCharacterNotFound(): void
    {
        $this->queryPort->method('getCharacter')->willReturn(null);
        $result = $this->controller->getCharacter(
            $this->authed('GET', '/api/match/m1/characters/c1'), $this->resp(),
            ['uuidMatch' => 'm1', 'uuidCharacter' => 'c1']
        );
        $this->assertSame(404, $result->getStatusCode());
    }
}
